File: src/main/java/com/anod/appwatcher/wishlist/WishlistFragment.java
package com.anod.appwatcher.wishlist;

import android.accounts.Account;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.android.volley.VolleyError;
import com.anod.appwatcher.R;
import com.anod.appwatcher.market.PlayStoreEndpoint;
import com.anod.appwatcher.market.WishlistEndpoint;
import com.anod.appwatcher.model.AppInfo;
import com.anod.appwatcher.model.AppInfoMetadata;
import com.anod.appwatcher.model.WatchAppList;
import com.anod.appwatcher.tags.TagSnackbar;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * @author algavris
 * @date 16/12/2016.
 */

public class WishlistFragment extends Fragment implements WatchAppList.Listener, PlayStoreEndpoint.Listener {
    public static final String TAG = "wishlist";
    public static final String EXTRA_ACCOUNT = "extra_account";
    public static final String EXTRA_AUTH_TOKEN = "extra_auth_token";

    @BindView(R.id.loading)
    LinearLayout mLoading;
    @BindView(android.R.id.list)
    RecyclerView mListView;
    @BindView(android.R.id.empty)
    TextView mEmptyView;
    @BindView(R.id.retry_box)
    LinearLayout mRetryView;
    @BindView(R.id.retry)
    Button mRetryButton;

    private WishlistEndpoint mEndpoint;
    private WatchAppList mWatchAppList;

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);

        if (mEndpoint == null) {
            mEndpoint = new WishlistEndpoint(context, true);
        }

        if (mWatchAppList == null)
        {
            mWatchAppList = new WatchAppList(this);
        }

        mWatchAppList.attach(context);
        mEndpoint.setListener(this);
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mEndpoint.setListener(null);
        mWatchAppList.detach();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_wishlist, container, false);
        ButterKnife.bind(this, view);

        Context context = getContext();

        mListView.setLayoutManager(new LinearLayoutManager(context));
        mRetryButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                mEndpoint.startAsync();
            }
        });

        mListView.setVisibility(View.GONE);
        mEmptyView.setVisibility(View.GONE);
        mLoading.setVisibility(View.VISIBLE);
        mRetryView.setVisibility(View.GONE);

        getActivity().setTitle(R.string.wishlist);

        return view;
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        Account account = getArguments().getParcelable(EXTRA_ACCOUNT);
        String authToken = getArguments().getString(EXTRA_AUTH_TOKEN);

        if (account == null || TextUtils.isEmpty(authToken)) {
            Toast.makeText(getContext(), R.string.choose_an_account, Toast.LENGTH_SHORT).show();
            getActivity().finish();
        } else {
            startLoadingList(account, authToken);
        }
    }

    @Override
    public void onWatchListChangeSuccess(AppInfo info, int newStatus) {
        if (newStatus == AppInfoMetadata.STATUS_NORMAL) {
            TagSnackbar.make(getActivity(), info, false).show();
        }
        mListView.getAdapter().notifyDataSetChanged();
    }

    @Override
    public void onWatchListChangeError(AppInfo info, int error) {
        if (WatchAppList.ERROR_ALREADY_ADDED == error) {
            Toast.makeText(getContext(), R.string.app_already_added, Toast.LENGTH_SHORT).show();
            mListView.getAdapter().notifyDataSetChanged();
        } else if (error == WatchAppList.ERROR_INSERT) {
            Toast.makeText(getContext(), R.string.error_insert_app, Toast.LENGTH_SHORT).show();
        }
    }


    private void startLoadingList(@NonNull Account account,@NonNull String authSubToken) {
        mEndpoint.setAccount(account, authSubToken);

        Context context = getContext();

        ResultsAdapterWishlist adapter = new ResultsAdapterWishlist(context, mEndpoint , mWatchAppList);
        mListView.setAdapter(adapter);

        mEndpoint.startAsync();
    }


    private void showRetryButton() {
        mListView.setVisibility(View.GONE);
        mEmptyView.setVisibility(View.GONE);
        mLoading.setVisibility(View.GONE);
        mRetryView.setVisibility(View.VISIBLE);
    }

    private void showListView() {
        mListView.setVisibility(View.VISIBLE);
        mEmptyView.setVisibility(View.GONE);
        mLoading.setVisibility(View.GONE);
        mRetryView.setVisibility(View.GONE);
    }

    private void showNoResults() {
        mLoading.setVisibility(View.GONE);
        mListView.setVisibility(View.GONE);
        mRetryView.setVisibility(View.GONE);
        mEmptyView.setText(R.string.no_result_found);
        mEmptyView.setVisibility(View.VISIBLE);
    }

    @Override
    public void onDataChanged() {
        if (mEndpoint.getData().getCount() == 0) {
            showNoResults();
        } else {
            showListView();
            mListView.getAdapter().notifyDataSetChanged();
        }
    }

    @Override
    public void onErrorResponse(VolleyError error) {
        mLoading.setVisibility(View.GONE);
        showRetryButton();
    }
}

File: app/src/main/java/com/anod/appwatcher/wishlist/WishlistFragment.kt
package com.anod.appwatcher.wishlist

import android.accounts.Account
import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.compose.ui.platform.ViewCompositionStrategy
import androidx.fragment.app.Fragment
import com.anod.appwatcher.compose.AppTheme
import com.anod.appwatcher.databinding.FragmentWishlistBinding
import info.anodsplace.framework.app.CustomThemeColors
import info.anodsplace.framework.app.FragmentContainerFactory
import info.anodsplace.framework.app.FragmentToolbarActivity
import org.koin.core.component.KoinComponent

/**
 * @author Alex Gavrishev
 * *
 * @date 16/12/2016.
 */
class WishListFragment : Fragment(), KoinComponent {
    private var _binding: FragmentWishlistBinding? = null
    private val binding get() = _binding!!


    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        _binding = FragmentWishlistBinding.inflate(inflater, container, false)
        binding.composeView.apply {
            // Dispose of the Composition when the view's LifecycleOwner
            // is destroyed
            setViewCompositionStrategy(ViewCompositionStrategy.DisposeOnViewTreeLifecycleDestroyed)
            setContent {
                // In Compose world
                AppTheme {

                }
            }

        }
        return binding.root
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
//
//    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
//        super.onViewCreated(view, savedInstanceState)
//
//        binding.list.layoutManager = LinearLayoutManager(context)
//        binding.retryButton.setOnClickListener {
//            load()
//        }
//
//        binding.list.visibility = View.GONE
//        binding.empty.visibility = View.GONE
//        binding.loading.visibility = View.VISIBLE
//        binding.retryView.visibility = View.GONE
//
//        requireActivity().setTitle(R.string.wishlist)
//
//        val account = requireArguments().getParcelable<Account>(EXTRA_ACCOUNT)
//        val authToken = requireArguments().getString(EXTRA_AUTH_TOKEN) ?: ""
//
//        if (account == null || authToken.isEmpty() || context == null) {
//            Toast.makeText(context, R.string.choose_an_account, Toast.LENGTH_SHORT).show()
//            requireActivity().finish()
//            return
//        } else {
//            this.adapter = ResultsAdapterList(requireContext(), action, viewModel.packages, iconLoader = get(), uploadDateParserCache = get())
//            binding.list.adapter = this.adapter
//        }
//
//        viewModel.appStatusChange.observe(viewLifecycleOwner) {
//            val newStatus = it.first
//            if (newStatus == AppInfoMetadata.STATUS_NORMAL) {
//                TagSnackbar.make(binding.root, it.second!!, false, requireActivity(), prefs).show()
//                binding.list.adapter!!.notifyDataSetChanged()
//            }
//        }
//
//        viewLifecycleOwner.lifecycleScope.launch {
//            viewModel.packages.collectLatest {
//                binding.list.adapter?.notifyDataSetChanged()
//            }
//        }
//
//        viewLifecycleOwner.lifecycleScope.launchWhenResumed {
//            action.collectLatest {
//                when (it) {
//                    is Delete -> viewModel.delete(it.info)
//                    is Add -> viewModel.add(it.info)
//                }
//            }
//        }
//
//        adapter.addLoadStateListener { loadStates ->
//            when (loadStates.refresh) {
//                is LoadState.Loading -> {
//                    binding.loading.isVisible = true
//                }
//                is LoadState.NotLoading -> {
//                    binding.loading.isVisible = false
//                    if (adapter.itemCount == 0) {
//                        showNoResults()
//                    } else {
//                        showListView()
//                    }
//                }
//                is LoadState.Error -> {
//                    binding.loading.isVisible = false
//                    showRetryButton()
//                }
//            }
//        }
//
//        load()
//    }
//
//    private fun load() {
//        loadJob?.cancel()
//        loadJob = lifecycleScope.launch {
//            viewModel.load().collectLatest {
//                adapter.submitData(it)
//            }
//        }
//    }
//
//    private fun showRetryButton() {
//        binding.list.visibility = View.GONE
//        binding.empty.visibility = View.GONE
//        binding.loading.visibility = View.GONE
//        binding.retryView.visibility = View.VISIBLE
//    }
//
//    private fun showListView() {
//        binding.list.visibility = View.VISIBLE
//        binding.empty.visibility = View.GONE
//        binding.loading.visibility = View.GONE
//        binding.retryView.visibility = View.GONE
//    }
//
//    private fun showNoResults() {
//        binding.loading.visibility = View.GONE
//        binding.list.visibility = View.GONE
//        binding.retryView.visibility = View.GONE
//        binding.empty.setText(R.string.no_result_wishlist)
//        binding.empty.visibility = View.VISIBLE
//    }

    companion object {
        const val EXTRA_ACCOUNT = "extra_account"
        const val EXTRA_AUTH_TOKEN = "extra_auth_token"

        private class Factory : FragmentContainerFactory("wishlist") {
            override fun create() = WishListFragment()
        }

        fun intent(context: Context, themeRes: Int, themeColors: CustomThemeColors, account: Account?, authToken: String?) = FragmentToolbarActivity.intent(
                context,
                Factory(),
                Bundle().apply {
                    putParcelable(EXTRA_ACCOUNT, account)
                    putString(EXTRA_AUTH_TOKEN, authToken)
                },
                themeRes,
                themeColors)
    }
}